datafusion/core/src/physical_plan/aggregates/row_hash.rs

```diff
@@ -219,91 +221,76 @@ impl GroupedHashAggregateStream {
-
+                            let result = self.group_aggregate_batch(batch);
                             timer.done();
 
                             // allocate memory
                             // This happens AFTER we actually used the memory, but simplifies the whole accounting and we are OK with
                             // overshooting a bit. Also this means we either store the whole record batch or not.
                             match result.and_then(|allocated| {
-                                this.row_aggr_state.reservation.try_grow(allocated)
+                                self.row_aggr_state.reservation.try_grow(allocated)
                             }) {
-                                Ok(_) => continue,
-                                Err(e) => Err(ArrowError::ExternalError(Box::new(e))),
+                                Ok(_) => {}
+                                Err(e) => {
+                                    return Poll::Ready(Some(Err(
+                                        ArrowError::ExternalError(Box::new(e)),
+                                    )))
+                                }
                             }
                         }
-                        Some(Err(e)) => Err(e),
+                        // inner had error, return to caller
+                        Some(Err(e)) => return Poll::Ready(Some(Err(e))),
+                        // inner is done, producing output
                         None => {
-                            let timer = this.baseline_metrics.elapsed_compute().timer();
-                            let result = create_batch_from_map(
-                                &this.mode,
-                                &this.row_converter,
-                                &this.row_aggr_schema,
-                                this.batch_size,
-                                this.row_group_skip_position,
-                                &mut this.row_aggr_state,
-                                &mut this.row_accumulators,
-                                &this.schema,
-                                &this.indices,
-                            );
-
-                            timer.done();
-                            result
+                            self.exec_state = ExecutionState::ProducingOutput;
                         }
-                    };
-
-                this.row_group_skip_position += this.batch_size;
-                return match result {
-                    Ok(Some(result)) => {
-                        let batch = result.record_output(&this.baseline_metrics);
-                        Some((Ok(batch), this))
                     }
-                    Ok(None) => None,
-                    Err(error) => Some((Err(error), this)),
-                };
-            }
-        });
+                }
 
-        // seems like some consumers call this stream even after it returned `None`, so let's fuse the stream.
-        let stream = stream.fuse();
```

Comment:
We used to fuse the stream implicitly -- but it is now handled via `ExecutionState::Done`